### PR TITLE
Replace MyGet with NuGet as default package source

### LINKED ISSUE #2261: Change default package source to nuget.org or All feed
As we complete the transition and release of the new editor package manager, it will be required signal more clearly the move to nuget.org, both on the default package manager feed and on the updates page when clicking the "Updates Available" button.

Currently the default feed is still preferred, with a couple of alternative options available:

- Switch to the "All" feed. This would make sense for the case of "Updates Available" mostly, but might be confusing when MyGet or other legacy feeds are considered which do not implement package type filtering (excess of packages might show up in this case).
- Switch the default to "nuget.org". This would be a nice default for browse, but might preclude other valid configurations.

Experimenting with a few prototypes might help gain an intuition for what might work best. It might also be worth exploring what the defaults for the VS package manager are, in case they have conventions which may inspire our decision.

### LINKED ISSUE #2271: Remove package license URL and package icon URL
With the move to NuGet we should stop using URLs for package license and package icons.

https://github.com/bonsai-rx/bonsai/blob/23f59395e23f9f4c81bcbea35030799b3d1f9889/tooling/Common.csproj.props#L25

### PR DESCRIPTION
**Pull request by glopesdev**
As detailed in #2281, we are moving away from MyGet as the official package feed. By leveraging the package type filtering features of the [SearchQueryService](https://learn.microsoft.com/en-us/nuget/api/search-query-service-resource) we do not require externally curated feeds.

We also replace package and icon URLs with embedded package content since NuGet supports the latest package metadata specification.

Fixes #2261 
Fixes #2271

### CODE REVIEW

File: .github/workflows/create-portable-zip.py [hunk 1]
```diff
@@ -24,8 +24,7 @@
         '<configuration>',
         '  <packageSources>',
         '    <add key="Gallery" value="Gallery" />',
-        '    <add key="Bonsai Packages" value="https://www.myget.org/F/bonsai/api/v3/index.json" />',
```
[the remark this answers is not in the document]
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py:27]: We could also keep them around but have them be disabled by default. (Maybe only in the installer? It'd be a bit obnoxious to clutter the `NuGet.config` in new Bonsai environment initialized from the portable zip.)

File: .github/workflows/create-portable-zip.py [hunk 2]
```diff
@@ -24,8 +24,7 @@
...
-        '    <add key="Community Packages" value="https://www.myget.org/F/bonsai-community/api/v3/index.json" />',
+        '    <add key="Bonsai Packages" value="https://api.nuget.org/v3/index.json" />',
```
Comment (PathogenDavid) [inline .github/workflows/create-portable-zip.py:27]: Mixed feelings on "rebranding" nuget.org like this.

This is going to result in "Bonsai Packages" and "nuget.org" package sources with no clear distinction (and in fact, there won't actually be any difference between them.)

It'll also cause `nuget.org` to queried for packages twice in situations where the package won't resolve. (Typically `Bonsai Packages` will end up being first priority and `nuget.org` will end up being last so in theory this is only a problem when the package won't resolve. It'd be quite annoying to a user experiencing poor internet connectivity though as it'd cause the failure to be delayed that much more.)

------------------------

I agree with the overall sentiment of labeling it as "Bonsai Packages" though.

Maybe we should just `<clear />` to eliminate machine-wide package sources by default? It's actually somewhat sensible that you would not want ambient state affecting Bonsai environments saved in a repository. (Downside would be the negative impact on people who configure a corporate nuget.org proxy, no idea if anyone like that uses Bonsai.)

Alternatively we could just disable the default `nuget.org` source, IE:

```xml
<disabledPackageSources>
  <add key="nuget.org" value="true" />
</disabledPackageSources>
```

This has the benefit of still showing `nuget.org` in the package source configuration dialog where it's clear that A) it's still there but disabled and B) makes it easily noticeable that it's redundant to `Bonsai Packages` so the user can guess why it's off:

Screenshot of the package sources dialog

In either case though `nuget.org` disappears from the package source dropdown by default. I know needing to explicitly select it for `nuget.org`-only packages has been a source of confusion in the past, which might mean having it disappear would result in similar confusion.

**Review by PathogenDavid — CHANGES_REQUESTED** [review]

File: .github/workflows/create-portable-zip.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py:27]: I was going to say that if we keep them around but marked as legacy that it'd be a good idea to automatically rename them, but we'd have to do it manually as the API provided by `IPackageSourceProvider` and `ISettings` isn't properly designed to allow renaming sources.

(The API should really have you provide an `(oldSource, newSource)` pair but it only accepts the `newSource` and tries to guess which one to update...using its name. What's really unfortunate is that the guts of doing the update [`SettingItem.Update`](https://github.com/NuGet/NuGet.Client/blob/e4e3b79701686199bc804a06533d2df054924d7e/src/NuGet.Core/NuGet.Configuration/Settings/SettingItem.cs#L42) actually could handle this properly, but it's only really used by [`SettingSection.Update`](https://github.com/NuGet/NuGet.Client/blob/e4e3b79701686199bc804a06533d2df054924d7e/src/NuGet.Core/NuGet.Configuration/Settings/SettingSection.cs#L46) which is where this API design flaw begins.)

File: Bonsai.Editor/GraphModel/UpgradeHelper.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace Bonsai.Editor.GraphModel
 {
     static class UpgradeHelper
     {
-        static readonly SemanticVersion DeprecationTarget = SemanticVersion.Parse("2.7.0");
+        static readonly SemanticVersion DeprecationTarget = SemanticVersion.Parse("2.9.0");
```
[the remark this answers is not in the document]
Comment (glopesdev) [reply Bonsai.Editor/GraphModel/UpgradeHelper.cs:14]: @PathogenDavid this makes the editor show up a warning for all workflows created with editor version < 2.9 warning that formats have changed and requesting explicit saving of old workflows, rather than assuming it is safe to overwrite the previous file.

We used this common practice whenever we implemented forward compatibility upgrades of the workflow to avoid having source files touched without anyone realizing it following an editor update.

To be fair, nothing in terms of the workflow itself has changed in this version. However, I feel there are sufficient changes regarding all the other editor-adjacent files (`.layout` files, `.editor` files, the new project folder structure, etc) to merit this warning, so that no-one is surprised when `.layout` files suddenly start disappearing, for example.

File: .github/workflows/create-portable-zip.py (hunk 2 above)
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py]: Ah, that makes way more sense!

> I was never quite sure how the system-wide `nuget.org` feed ends up in the VS package manager (not sure if it is "installed" by VS with this standard name, or by something else).

The NuGet client API adds it automagically. Specifically it happens in [`NuGet.Configuration.FileSystemUtility.GetOrCreateDocument`](https://github.com/NuGet/NuGet.Client/blob/c9176652a31d04995f11131a60ebfd9ac90ecaa9/src/NuGet.Core/NuGet.Configuration/Utility/FileSystemUtility.cs#L16-L29), which is called to [load](https://github.com/NuGet/NuGet.Client/blob/e4e3b79701686199bc804a06533d2df054924d7e/src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs#L119) any and all `NuGet.config` files used by NuGet. As the name implies, `GetOrCreateDocument` will create any files that don't yet exist.

From a Bonsai perspective, this happens when [we call](https://github.com/bonsai-rx/bonsai/blob/44253e33b5d521fd8ea0fddc26f7836cf6f15aef/Bonsai.Configuration/Bootstrapper.cs#L23) `NuGet.Configuration.Settings.LoadDefaultSettings`.

So it's not really necessary to include it at all if we don't want to. Unless someone went out of their way to remove the package source it will always be there.

<details>
<summary>Full stack trace for curious nerds (click to expand)</summary>

Here's the full stack trace with minor substitutions to make it easier to read:

```
NuGet.Configuration.dll!NuGet.Configuration.FileSystemUtility.GetOrCreateDocument(content = {System.Xml.Linq.XDocument}, fullPath = "/AppData/Roaming/NuGet/NuGet.Config") Line 18
NuGet.Configuration.dll!NuGet.Configuration.SettingsFile..ctor.AnonymousMethod__0() Line 114
NuGet.Configuration.dll!NuGet.Configuration.SettingsFile.ExecuteSynchronized.AnonymousMethod__0() Line 200
NuGet.Common.dll!NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLocked(filePath = "/AppData/Roaming/NuGet/NuGet.Config", action = {AnonymousMethod__0}) Line 190
NuGet.Configuration.dll!NuGet.Configuration.SettingsFile.ExecuteSynchronized(ioOperation = {AnonymousMethod__0}) Line 196
NuGet.Configuration.dll!NuGet.Configuration.SettingsFile.SettingsFile(directoryPath = "/AppData/Roaming/NuGet", fileName = "NuGet.Config", isMachineWide = false, isReadOnly = false) Line 112
NuGet.Configuration.dll!NuGet.Configuration.Settings.ReadSettings(settingsRoot = "/path/to/.bonsai", settingsPath = "/AppData/Roaming/NuGet/NuGet.Config", isMachineWideSettings = false, isAdditionalUserWideConfig = false, settingsLoadingContext = null) Line 782
NuGet.Configuration.dll!NuGet.Configuration.Settings.LoadUserSpecificSettings(root = "/path/to/.bonsai", configFileName = null, useTestingGlobalPath = false, settingsLoadingContext = null) Line 551
mscorlib.dll!System.Collections.Generic.List<NuGet.Configuration.SettingsFile>.InsertRange(index = 3, collection) Line 1244
NuGet.Configuration.dll!NuGet.Configuration.Settings.LoadSettingsForSpecificConfigs(root = "/path/to/.bonsai", configFileName = null, validSettingFiles = { Count = 3 }, machineWideSettings = {BonsaiMachineWideSettings}, loadUserWideSettings = true, useTestingGlobalPath = false, settingsLoadingContext = null) Line 494
NuGet.Configuration.dll!NuGet.Configuration.Settings.LoadSettings(string root = "/path/to/.bonsai", configFileName = null, machineWideSettings = {BonsaiMachineWideSettings}, loadUserWideSettings = true, useTestingGlobalPath = false, settingsLoadingContext = null) Line 440
NuGet.Configuration.dll!NuGet.Configuration.Settings.LoadDefaultSettings(root = "/path/to/.bonsai", configFileName = null, machineWideSettings = {BonsaiMachineWideSettings}, settingsLoadingContext = null) Line 337
NuGet.Configuration.dll!NuGet.Configuration.Settings.LoadDefaultSettings(root = "/path/to/.bonsai", configFileName = null, machineWideSettings = {BonsaiMachineWideSettings}) Line 297
Bonsai.Configuration.dll!Bonsai.Configuration.Bootstrapper.Bootstrapper(path = "/path/to/.bonsai/Packages") Line 23
Bonsai.exe!Bonsai.EditorBootstrapper.EditorBootstrapper(path = "/path/to/.bonsai/Packages") Line 29
Bonsai.exe!Bonsai.Program.Main(args = {}) Line 176
```

</details>

-----------------

Of course the proof is in the pudding. Here's a fresh Windows 10 install where I've literally done nothing but launch Bonsai once from the portable zip:

image

(I included the AppData folder in the screenshot to emphasize how barren this installation of Windows truly is.)

[Windows Sandbox](https://learn.microsoft.com/en-us/windows/security/application-security/application-isolation/windows-sandbox/) is very handy for testing these sorts of things BTW. It's how I test changes to the Bonsai installer. (Basically gives you a 0-effort completely fresh vanilla[^0] Windows VM with one button click in like 5 seconds.)

Just run this command from an elevated PowerShell to enable it and find it in the start menu:

```pwsh
Enable-WindowsOptionalFeature -FeatureName "Containers-DisposableClientVM" -All -Online
```

[^0]: It's not *totally* vanilla. If has some unnecessary stuff disabled for performance reasons. For example, Windows Update, Windows Defender, and search indexing are all disabled.

File: Bonsai.Editor/GraphModel/UpgradeHelper.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (PathogenDavid) [reply Bonsai.Editor/GraphModel/UpgradeHelper.cs:14]: Oh that makes sense, I was too focused on trying to figure out how it was related to NuGet things. I should've read the commit message 😅

File: .github/workflows/create-portable-zip.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py:27]: > This is why I started feeling we should just remove the entire band-aid in one go.

Makes sense, I'm here for it.

> since usually they are about reproducibility most of all.

True, but upgrading your Bonsai version is throwing reproducibility out the window since you've changed something fundamental.

> The `NuGet.config` file might even contained new sources which have been added manually since it is much more visible in the bootstrapper folder.

Yeah, I definitely wouldn't just blow it away. I was imagining calling `IPackageSourceProvider.RemovePackageSource` to remove the specific sources.

[the remark this answers is not in the document]
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py:27]: As an aside, when when getting that screenshot in my other comment I was reminded of how absurdly slow Bonsai restores when the MyGet package sources are enabled 😭

If you hadn't already convinced me that alone would be reason enough to cast them into the fire

**Review by PathogenDavid — APPROVED** [review]
LGTM unless you want to remove the technically-unnecessary `nuget.org` package source and/or I indirectly convinced you to proactively purge the legacy sources.

File: .github/workflows/create-portable-zip.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply .github/workflows/create-portable-zip.py:27]: > Yeah, I definitely wouldn't just blow it away. I was imagining calling `IPackageSourceProvider.RemovePackageSource` to remove the specific sources.

I wonder whether this would be impacted by the point you raised elsewhere of the package source provider preferring to modify global settings rather than local? What you propose makes sense, but I would like to make sure there is a way to avoid destroying shared settings accidentally.

File: .github/workflows/create-portable-zip.py (hunk 2 above)
Comment (glopesdev) [reply .github/workflows/create-portable-zip.py]: > The NuGet client API adds it automagically.

Ah, that is interesting I don't think their client did that in earlier versions, as I still remember having to add `nuget.org` by hand in the settings editor.

In that case I am almost tempted to simply remove the [Bonsai machine-wide settings config file](https://github.com/bonsai-rx/bonsai/blob/44253e33b5d521fd8ea0fddc26f7836cf6f15aef/Bonsai.Setup/Bonsai.config) entirely. The only other entry in there was for the `Gallery` folder but it's not really used most of the time, as there are better ways to setup a local package source for testing. Initially it was intended to "cache" gallery packages, but I don't think this makes sense with the NuGet package cache in place, might be easier to simply use the packages directly from wherever they are being stored.

> [Windows Sandbox](https://learn.microsoft.com/en-us/windows/security/application-security/application-isolation/windows-sandbox/) is very handy for testing these sorts of things BTW.

Oh, I have been wanting something like this for a long time, thanks for bringing it up! I guess it is one of the natural advantages of everything running inside Hypervisor anymore, I guess there is no such thing as a "native windows" anymore, everything is a VM.

**Comment by glopesdev** [comment]
> LGTM unless you want to remove the technically-unnecessary nuget.org package source and/or I indirectly convinced you to proactively purge the legacy sources.

@PathogenDavid I think I am convinced to do both. I have implemented the former in the last commit if you want to take a look or test for edge cases. For the latter I am not sure how best to ensure we purge legacy sources only from the closest `NuGet.config` in the case of environments. If there is no good way perhaps we can leave them a while longer?

**Comment by glopesdev** [comment]
Another bit to consider is whether we still need this:

https://github.com/bonsai-rx/bonsai/blob/44253e33b5d521fd8ea0fddc26f7836cf6f15aef/Bonsai.Configuration/ScriptExtensions.cs#L32

It was used when exporting the project file `Extensions.csproj` for local extensions, since once upon a time we would not be able to restore specific packages from VS code if they were not on `nuget.org`. If we assume now everything must come from NuGet, perhaps it makes sense to simply drop this behavior.

I guess it might do it anyway if there were no package sources in the local machine-wide settings, e.g. a Bonsai environment, but if we do not aim to rely on this again, perhaps might as well remove it?

File: .github/workflows/create-portable-zip.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py:27]: It's not quite the same situation, but you're right. There's a similar problem there.

In that case I'd probably just modify the `NuGet.config` manually (similar to the `Extensions.csproj` handling.)

We could also try to passively inform people when they have the stale sources around, maybe an alert next to the package sources dropdown or the source config button or something. However I think for now it's better to stop holding up this PR and Bonsai 2.9 with bike shedding so I'm good with leaving things as-is 😅

File: .github/workflows/create-portable-zip.py (hunk 2 above)
Comment (PathogenDavid) [reply .github/workflows/create-portable-zip.py]: > In that case I am almost tempted to simply remove the entire [Bonsai machine-wide settings config file](https://github.com/bonsai-rx/bonsai/blob/44253e33b5d521fd8ea0fddc26f7836cf6f15aef/Bonsai.Setup/Bonsai.config).

I actually almost suggested this before remembering the `Gallery` source exists. If you don't think it's valuable I'm definitely OK with removing it. We could drop the `NuGet.config` from Bonsai environments entirely.

(Or maybe the NuGet.config used for Bonsai environments should always `<clear />` everything and start fresh in the name of reproducibility 😅.)

> EDIT: I guess there was also an entry for the "Unstable" packages coming from NuGet.org, so I guess perhaps we could leave this for the time being.

Ah yeah, I kind-of forgot about that too.

We should just ship an empty NuGet.config for now. In theory it might be nice for https://github.com/bonsai-rx/bonsai/issues/2294 if we pursue that and decide that it should prefer updating the "nearest" NuGet.config (which has its own quirks that'd need to be discussed.)

Plus having the empty `NuGet.config` for now ensures that anyone who updates a Bonsai environment by manually copying files out of a portable zip is less likely to accidentally keep the old package sources around.

The `<clear />` idea is probably worth discussing in dev club too.

**Comment by PathogenDavid** [comment]
> I think I am convinced to do both.

Ah I should've looked for top-level comments before replying to the review ones 😅

>  I have implemented the former in the last commit if you want to take a look or test for edge cases.

> For the latter I am not sure how best to ensure we purge legacy sources only from the closest `NuGet.config` in the case of environments.

Looks good to me! I pulled the installed from CI and tested installing it on its own as well as using it to upgrade from Bonsai 2.8.5. The machine-wide config was automatically removed as expected in the latter case.

One minor thing is that the `Gallery` folder still gets created. Should it be removed from the installer too? (Not sure if that'd cause it to be deleted on upgrade. We probably don't want to blow away any files in there but we probably don't want to keep adding it either.)

> [`EnsureNuGetSettings` ...], perhaps might as well remove it?

I think this is safe to remove.

In theory something like this would be necessary in some edge cases, but it doesn't appear to actually be covering those edge cases right now anyway. (EG: `Extensions.csproj` would not see a package source coming from `.bonsai/NuGet.config` - you'd have to put it in the root `NuGet.config`.)

I think we just file those edge cases under the "`Extensions.csproj` is not a full-blown C# project and you should not do anything strange with it" umbrella.[^ramble]

[^ramble]: If we *did* want to solve that issue, I'd use a much different approach than populating an automatically generated `NuGet.config`. We could instead set the `RestoreConfigFile` property as appropriate. It is not [documented](https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-properties) as such, but I checked [the implementation](https://github.com/NuGet/NuGet.Client/blob/a9234756381d6656ae9463def7bafddc5a8cb76e/src/NuGet.Core/NuGet.Build.Tasks/RestoreSettingsUtils.cs#L55-L62) and there is an implicit behavior of walking up the directory tree if you specify the property as `C:\A\Path\Ending\In\A\Backslash\`. Although it's probably not intentional that it works that way, it actually breaks when `RestoreProjectStyle` is the legacy `PackagesConfig`. `RestoreRootConfigDirectory` is a more explicit property for this purpose but is technically undocumented.)

File: .github/workflows/create-portable-zip.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply .github/workflows/create-portable-zip.py:27]: Sounds good, I have removed all sources from the packaged `NuGet.config` file so it will stop proliferating. We can take a more active stance in cleaning up later if it becomes a problem.

**Comment by glopesdev** [comment]
> One minor thing is that the Gallery folder still gets created. Should it be removed from the installer too?

Yup, just did in https://github.com/bonsai-rx/bonsai/pull/2285/commits/418bfa9fa8460478609d545e4c7a856a470380ed. I think this one should be good to go now, merging soon.
